File: src/leetCode/c++/Solution101.cpp
//DFS O(V) O(H)
class Solution101 {
public:
    bool isSymmetric(TreeNode* root) {
        if(!root) return true;
        return check(root->left, root->right);
    }

    bool check(TreeNode* a, TreeNode* b){
        if(!a || !b) return !a && !b;
        if(a->val != b->val) return false;
        return check(a->left, b->right) && check(a->right, b->left);
    }
};

File: src/leetCode/java/Solution101.java
package leetCode.java;

//DFS O(V) O(H)
class Solution101 {
    public boolean isSymmetric(TreeNode root) {
        if(root == null) return true;
        return check(root.left, root.right);
    }

    public boolean check(TreeNode a, TreeNode b){
        if(a == null || b == null) return a == null && b == null;
        if(a.val != b.val) return false;
        return check(a.left, b.right) && check(a.right, b.left);
    }
}